In [215]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy as scipy
import matplotlib.pyplot as plt
import plotly as px
from datetime import timedelta

###### Выгрузим наши таблицы по ссылке, проверим размер, типы данных

In [261]:
customers = pd.read_csv('https://getfile.dokpub.com/yandex/get/https://disk.yandex.ru/d/QXrVDUoQvFhIcg')
orders = pd.read_csv('https://getfile.dokpub.com/yandex/get/https://disk.yandex.ru/d/0zWmh1bYP6REbw')
items = pd.read_csv('https://getfile.dokpub.com/yandex/get/https://disk.yandex.ru/d/xCsQ0FiC-Ue4Lg')

In [262]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [263]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [264]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [265]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [266]:
items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [267]:
items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


##### Преобразуем типы данных в колонках со временем заказов во временные

In [268]:
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'])
orders['order_approved_at'] = pd.to_datetime(orders['order_approved_at'])
orders['order_delivered_carrier_date'] = pd.to_datetime(orders['order_delivered_carrier_date'])
orders['order_delivered_customer_date'] = pd.to_datetime(orders['order_delivered_customer_date'])
orders['order_estimated_delivery_date'] = pd.to_datetime(orders['order_estimated_delivery_date'])
items['shipping_limit_date'] = pd.to_datetime(items['shipping_limit_date'])

##### Объединим наши данные в одну таблицу для удобства работы с ними

In [269]:
customers_orders = customers.merge(orders, on = 'customer_id')

In [270]:
customers_orders.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  
0           2017-05-25 10:35:35                    2017-06-05  
1           2018-01-29 12:41:19                    2018-02-06  
2           2018-06-14 17:58:51                    2018-06-13  
3           2018-03-28 16:04:25                    2018-04-10  
4           2018-08-09 20:55:48                    2018-08-15

In [271]:
# проверим корректность объединения - все строки скопированы.
customers_orders.shape

(99441, 12)

### 1. Сколько у нас пользователей, которые совершили покупку только один раз?

Проверим, какие статусы могут принимать наши заказы.

In [272]:
customers_orders.order_status.value_counts()

delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: order_status, dtype: int64

Проверим нашу таблицу на наличие нулевых значений и рассмотрим их подробнее. 
Заказы, для которых нет значений в столбцах время передачи заказа в логистическую службу и время доставки заказа, мы пока не рассматриваем. Рассмотрим заказы, по которым нет подтверждения оплаты (order_approved_at)

In [273]:
customers_orders.isnull().sum()

customer_id                         0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
order_id                            0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Видим, что есть заказы, для которых не подтверждена оплата, но которые либо только созданы (что мы еще не будем считать покупкой), либо которые уже доставлены покупателю (возможно, с какими-то клиентами заключены особые договорные отношения и для них есть отсрочка платежа, либо они оплачивали заказ наличными курьеру, либо данные в системе не прогрузились, следовательно такие заказы все равно являются покупками)

In [274]:
customers_orders[customers_orders[['order_approved_at']].isna().any(axis=1)]\
    .query("order_status != 'canceled'")\

customer_id                customer_unique_id  \
8970   0bf35cac6cc7327065da879e2d90fae8  c4c0011e639bdbcf26059ddc38bd3c18   
13047  4ee64f4bfc542546f422da0aeb462853  10d5890ed5e7ec78a7db23b5ac579b17   
13890  1e101e0daffaddce8159d25a8e53f2b2  c8822fce1d0bfa7ddf0da24fff947172   
20863  d5de688c321096d15508faae67a27051  d49f3dae6bad25d05160fc17aca5942d   
22703  07a2a7e0f63fd8cb757ed77d4245623c  79af1bbf230a2630487975aa5d7d6220   
27441  68d081753ad4fe22fc4d410a9eb1ca01  2e0a2166aa23da2472c6a60c4af6f7a6   
39286  d85919cb3c0529589c6fa617f5f43281  c094ac95fcd52f821809ec232a7a6956   
41493  74bebaf46603f9340e3b50c6b086f992  f79be7c08dd24b72d34634f1b89333a4   
44050  684cb238dc5b5d6366244e0e0776b450  6ff8b0d7b35d5c945633b8d60165691b   
45313  7d61b9f4f216052ba664f22e9c504ef1  161e5b2b4513edb336986a8909218365   
51094  964a6df3d9bdf60fe3e7b8bb69ed893a  2e015e68fbfa36910263df5384e8e45c   
52849  a3d3c38e58b9d2dfb9207cab690b6310  5a4fa4919cbf2b049e72be460a380e5b   
60039  2127dc6603ac33544953ef05ec155771  8a9a08c7ca8900a200d83cf838a07e0b   
61049  f67cd1a215aae2a1074638bbd35a223a  bc1896dc77f49e6dec880445a9b443a3   
71548  4c1ccc74e00993733742a3c786dc3c1f  91efb7fcabc17925099dced52435837f   
79266  438449d4af8980d107bf04571413a8e7  55fbb57fd54c35a38c80c2a032a6ca6e   
79535  29c35fc91fc13fb5073c8f30505d860d  7e1a5ca61b572d76b64b6688b9f96473   
85378  2941af76d38100e0f8740a374f1a5dc3  e1f01a1bd6485e58ad3c769a5427d8a8   
87242  725e9c75605414b21fd8c8d5a1c2f1d6  355862aa14375fe3950ee04fef9c7a53   

       customer_zip_code_prefix         customer_city customer_state  \
8970                      13225       varzea paulista             SP   
13047                     93025          sao leopoldo             RS   
13890                     27945                 macae             RJ   
20863                     36400  conselheiro lafaiete             MG   
22703                     37660          paraisopolis             MG   
27441                      3573             sao paulo             SP   
39286                     95795         sao vendelino             RS   
41493                     65110   sao jose de ribamar             MA   
44050                     11030                santos             SP   
45313                     87720             paranavai             PR   
51094                     73401              brasilia             DF   
52849                     35620                abaete             MG   
60039                      6708                 cotia             SP   
61049                     21020        rio de janeiro             RJ   
71548                     93548         novo hamburgo             RS   
79266                     15300       general salgado             SP   
79535                     62700               caninde             CE   
85378                      8230             sao paulo             SP   
87242                     22621        rio de janeiro             RJ   

                               order_id order_status order_purchase_timestamp  \
8970   d77031d6a3c8a52f019764e68f211c69    delivered      2017-02-18 11:04:19   
13047  35de4050331c6c644cddc86f4f2d0d64      created      2017-12-05 01:07:58   
13890  12a95a3c06dbaec84bcfb0e2da5d228a    delivered      2017-02-17 13:05:55   
20863  7002a78c79c519ac54022d4f8a65e6e8    delivered      2017-01-19 22:26:59   
22703  51eb2eebd5d76a24625b31c33dd41449    delivered      2017-02-18 15:52:27   
27441  d69e5d356402adc8cf17e08b5033acfb    delivered      2017-02-19 01:28:47   
39286  3c0b8706b065f9919d0505d3b3343881    delivered      2017-02-17 15:53:27   
41493  2babbb4b15e6d2dfe95e2de765c97bce    delivered      2017-02-18 17:15:03   
44050  c1d4211b3dae76144deccd6c74144a88    delivered      2017-01-19 12:48:08   
45313  90ab3e7d52544ec7bc3363c82689965f      created      2017-11-06 13:12:34   
51094  dba5062fbda3af4fb6c33b1e040ca38f      created      2018-02-09 17:21:04   
52849  2eecb0d85f281280f79fa00f9cec1a95    delivered      2017-02-17 17:21

Чтобы понять, будем ли мы считать покупкой заказы в статусе Invoiced, рассмотрим их подробнее. 

Мы видим, что для всех заказов, которым только выставлен счет, имеются данные о подтверждении оплаты, но нет никаких изменений в самом статусе заказа и дата доставки не назначены, хотя счет был выставлен существенно давно. Поэтому примем эти данные за ошибку выгрузки и предположим, что оплаты не было, иначе статус заказа был бы изменен.

Следовательно эти заказы мы покупкой НЕ считаем

In [275]:
customers_orders.query("order_status == 'invoiced'")

customer_id                customer_unique_id  \
47     c132855c926907970dcf6f2bf0b33a24  a8ae36a2bb6c2bbc3b5d62ede131c9ef   
133    faafae37ddb3e4af81c10a397bcca35f  1a7a10e9273fca58179bdd0d40843b0c   
392    1aa414b77e655c44696ea1d3e61d7ec7  cdbeeb59bf23bb27eba2b6a8dd4b9e09   
446    3d14ef1b615443e8569574ccf4c70ffb  57d2e3a4c3b277bea367027f105f1c02   
545    41b200d1ce8675f154c91c2da887bcee  9b05b38d7d9ef19a89815c217b30fd37   
...                                 ...                               ...   
98453  da5a76cb142d1caffc89b063aa35d07b  cc7e6b37d88a3469c01483f43b6ff947   
98546  0ae2b2edee0a604c92ccf6ed3c64376c  dea3bbce56fc4ca0aae5b3bdb784815b   
98815  56e8cb7692d76a84812cb3c93e1bcaf3  81ba91c294c42c45926e79a6527b8c66   
99087  99a9947c5c954fa3ea64dd9114e0f532  acdb729157837a904d393ed571aa60cd   
99176  da231380834f900e3fc3a958a1533a33  e3626c985901df173504f3d285ac823f   

       customer_zip_code_prefix   customer_city customer_state  \
47                        18740     taquarituba             SP   
133                        7243       guarulhos             SP   
392                       39404   montes claros             MG   
446                       83209       paranagua             PR   
545                        2114       sao paulo             SP   
...                         ...             ...            ...   
98453                      7041       guarulhos             SP   
98546                     89251  jaragua do sul             SC   
98815                     98801    santo angelo             RS   
99087                      4777       sao paulo             SP   
99176                     24937          marica             RJ   

                               order_id order_status order_purchase_timestamp  \
47     b5172d4a722c0ebb34c29bbbb7a3cb42     invoiced      2018-04-17 13:14:13   
133    e9376d90b28f9cdea83d4d850e4bb1a7     invoiced      2017-09-01 15:02:24   
392    899b8d92e1bdaece69745ffcc3c47d08     invoiced      2018-05-12 18:49:56   
446    f0cb299a9702d7db4e501895d61d2baf     invoiced      2017-06-10 22:46:28   
545    9dfd8f2ac479e34a16d68ff827759a23     invoiced      2018-05-15 15:55:01   
...                                 ...          ...                      ...   
98453  39d90234aa561e2f6a0ca08926c4f678     invoiced      2018-05-16 14:57:49   
98546  10605be2fc9784c0fd19b78f8b1924c0     invoiced      2017-08-26 10:06:15   
98815  8a43a2a954395f1f2911a8280575ab9c     invoiced      2017-06-06 19:26:58   
99087  823b6da4cafa14bb8a5374d408ce0fbf     invoiced      2018-05-16 16:04:24   
99176  42c5091238dc63339c79f5f35f8a5a6b     invoiced      2016-10-10 09:57:05   

        order_approved_at order_delivered_carrier_date  \
47    2018-04-18 05:31:51                          NaT   
133   2017-09-01 15:15:30                          NaT   
392   2018-05-12 19:16:36                          NaT   
446   2017-06-10 23:02:39                          NaT   
545   2018-05-17 02:15:28                          NaT   
...                   ...                          ...   
98453 2018-05-17 02:37:24                          NaT   
98546 2017-08-26 10:24:34                          NaT   
98815 2017-06-06 19:42:12                          NaT   
99087 2018-05-16 16:38:46                          NaT   
99176 2016-10-10 11:04:04                          NaT   

      order_delivered_customer_date order_estimated_delivery_date  
47                              NaT                    2018-05-24  
133                             NaT                    2017-09-15  
392                             NaT                    2018-06-01  
446                             NaT                    2017-07-03  
545                             NaT                    2018-06-01  
...                             ...                           ...  
98453                           NaT                    2018-06-04  
98546                           NaT                    2017-09-18  
98815                       

Изучив более внимательно данные по колонке order_approved_at, есть основания полагать, что данные об оплате выгружены некорректно, т.к. по логике продаж не должно быть подтверждения оплаты у заказов, где счет только выставлен, и наоборот - должно быть подтверждение у доставленных заказов. Т.к. у нас нет возможности точно определить, с чем связан данный факт и является ли он ошибкой, то для достоверности мы будем считать покупкой все заказы со статусом delivered.
Тем более мы предполагаем, что заказы, которые только созданы, либо товар по которым еще находятся в пути, могут быть отменены пользователем

Посчитаем количество уникальных пользователей, совершивших одну покупку

In [276]:
real_orders = customers_orders\
    .query("order_status == 'delivered'")\
    .groupby('customer_unique_id')\
    .agg({'order_id':'count'})\
    .query('order_id == 1').sum()

In [277]:
real_orders

order_id    90557
dtype: int64

##### Таким образом 90557 покупателей совершили только одну покупку. Учитывая общее количество заказов, можем сделать вывод, что более 90% покупателей к нам не возвращаются. Эти данные требуют более тщательного изучения, в чем причины такого большого процента нелояльных покупателей.

### 2. Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)?

##### Недоставленными заказами будем считать заказы, для которых отсутствует дата в колонке "время доставки заказа"

In [278]:
not_delivered = orders[['order_id', 'order_status', 'order_purchase_timestamp', 'order_delivered_customer_date']]
not_delivered = not_delivered[not_delivered.order_delivered_customer_date.isnull()]

In [279]:
not_delivered['month_year'] = not_delivered['order_purchase_timestamp'].dt.strftime('%Y-%m')

In [280]:
not_delivered = not_delivered\
    .groupby(['month_year','order_status'], as_index = False) \
    .agg({'order_id':'count'})\
    .pivot(columns = 'order_status', index = 'month_year', values = 'order_id')

In [44]:
not_delivered.style.highlight_max()

In [281]:
not_delivered.mean()

order_status
approved        1.000000
canceled       25.791667
created         1.666667
delivered       2.000000
invoiced       14.952381
processing     15.842105
shipped        48.130435
unavailable    29.000000
dtype: float64

##### Видим, что в среднем в месяц:
##### - 48 заказов не доставляется по причине того, что товар находится еще в пути (и мы ожидаем, что он будет все-таки доставлен), поэтому данный показатель для нас не очень интересен
##### - 29 заказов со статусом "недоступен". У нас нет данных, что подразуемвается под этим статусом. Возможно, закончился товар на складе и нашему отделу закупок стоит лучше прогнозировать товарные остатки:)
##### - 25 заказов отменены. Тут уже можно разбираться с причиной - возможно, покупателя не устроила предполагаемая дата доставки, либо другой магазин запустил какую-то интересную акцию на наш товар и покупатель ушел к конкурентам.

### 3. По каждому товару определить, в какой день недели товар чаще всего покупается.

Объединим наши данные с items

In [282]:
items_orders = customers_orders.merge(items, on = 'order_id', how='left')

Проверим, все ли присоединилось

In [283]:
items_orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 113425 entries, 0 to 113424
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   customer_id                    113425 non-null  object        
 1   customer_unique_id             113425 non-null  object        
 2   customer_zip_code_prefix       113425 non-null  int64         
 3   customer_city                  113425 non-null  object        
 4   customer_state                 113425 non-null  object        
 5   order_id                       113425 non-null  object        
 6   order_status                   113425 non-null  object        
 7   order_purchase_timestamp       113425 non-null  datetime64[ns]
 8   order_approved_at              113264 non-null  datetime64[ns]
 9   order_delivered_carrier_date   111457 non-null  datetime64[ns]
 10  order_delivered_customer_date  110196 non-null  datetime64[ns]
 11  

Добавим колонку с днем недели создания наших заказов

In [284]:
items_orders['weekday'] = items_orders['order_purchase_timestamp'].dt.day_name()

In [285]:
items_orders.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date  order_item_id  \
0           2017-05-25 10:35:35                    2017-06-05            1.0   
1           2018-01-29 12:41:19                    2018-02-06            1.0   
2           2018-06-14 17:58:51                    2018-06-13            1.0   
3           2018-03-28 16:04:25                    2018-04-10            1.0   
4           2018-08-09 20:55:48                    2018-08-15            1.0   

                         product_id                         seller_id  \
0  a9516a079e37a9c9c36b9b78b10169e8  7c67e1448b00f6e969d365cea6b010ab   
1  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
2  bd07b66896d6f1494f5b86251848ced7  7c67e1448b00f6e969d365cea6b010ab   
3  a5647c44af977b148e0a3a4751a09e2e  7c67e1448b00f6e969d365cea6b010ab   
4  9391a573abe00141c56e38d84d7d5b3b  4a3ca9315b744ce9f8e9374361493884   

  shipping_limit_date   price  freight_value   weekday  
0 2017-05-22 15:22:12  124.99          21.88   Tuesday  
1 2018-01-18 20:58:32  289.00          46.48    Friday  
2 2018-06-05 16:19:10  139.94          17.79  Saturday  
3 2018-03-27 16:31:16  149.94          23.36   Tuesday  
4 2018-07-31 10:10:09  230.00          22.25    Sunday

In [286]:
items_by_weekdays = items_orders[['product_id', 'order_purchase_timestamp', 'weekday']]

In [287]:
items_by_weekdays

product_id order_purchase_timestamp    weekday
0       a9516a079e37a9c9c36b9b78b10169e8      2017-05-16 15:05:35    Tuesday
1       4aa6014eceb682077f9dc4bffebc05b0      2018-01-12 20:48:24     Friday
2       bd07b66896d6f1494f5b86251848ced7      2018-05-19 16:07:45   Saturday
3       a5647c44af977b148e0a3a4751a09e2e      2018-03-13 16:06:38    Tuesday
4       9391a573abe00141c56e38d84d7d5b3b      2018-07-29 09:51:30     Sunday
...                                  ...                      ...        ...
113420  ccb4503d9d43d245d3b295d0544f988b      2018-04-07 15:48:17   Saturday
113421  9ede6b0570a75a4b9de4f383329f99ee      2018-04-04 08:20:22  Wednesday
113422  7a5d2e1e131a860ae7d18f6fffa9d689      2018-04-08 20:11:50     Sunday
113423  f819f0c84a64f02d3a5606ca95edd272      2017-11-03 21:08:33     Friday
113424  017692475c1c954ff597feda05131d73      2017-12-19 14:27:23    Tuesday

[113425 rows x 3 columns]

Посчитаем для наших товаров количество заказов с разбивкой по дням недели и выберем день с максимальным количеством заказов
Т.к. некоторые товары у нас в таблице повторяются несколько раз, т.к. были заказаны в разные дни недели, для каждого продукта оставим только день недели с максимальным количеством заказанного товара

In [288]:
items_by_weekdays = items_by_weekdays \
    .groupby(['weekday', 'product_id'], as_index = False)\
    .agg({'order_purchase_timestamp' : 'count'})\
    .rename(columns={'order_purchase_timestamp': 'quantity'})\
    .sort_values('quantity',ascending=False)\
    .drop_duplicates('product_id',keep='first')

##### Смотрим 10 наших наиболее заказываемых товаров (в один день недели)

In [289]:
items_by_weekdays.head(10)

weekday                        product_id  quantity
54190  Wednesday  422879e10f46682990de24d770e7f83d        93
14535     Monday  99a4788cb24856965c36a24e339b6058        92
39349   Thursday  aca2eb7d00ea1a7b8ebd4e68314663af        89
45392    Tuesday  53b36df67ebb7c41585e8d54d6772e08        76
35327   Thursday  389d119b48cf3043d311335e499d9c6b        67
1895      Friday  368c6c730842d78016ad823897a372db        67
54840  Wednesday  53759a2ecddad2bb87a079a1f1519f73        66
16669     Monday  d1c427060a0f73f6b889a5c7c61f2ac4        64
54013  Wednesday  3dd2a17168ec895c781a9191c1e95ad7        54
39129   Thursday  a62e25e09e05e6faf31d90c6ec1aa3d1        50

### 4. Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? 

##### Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть.

Оставим в таблице только интересующие нас колонки, а так же помним, что за покупку мы считаем заказ в статусе delivered

In [89]:
customers_orders.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date month_year  
0           2017-05-25 10:35:35                    2017-06-05    2017-05  
1           2018-01-29 12:41:19                    2018-02-06    2018-01  
2           2018-06-14 17:58:51                    2018-06-13    2018-05  
3           2018-03-28 16:04:25                    2018-04-10    2018-03  
4           2018-08-09 20:55:48                    2018-08-15    2018-07

In [90]:
customers_with_orders = customers_orders[['customer_unique_id', 'order_id', 'order_status','order_purchase_timestamp']]\
    .query('order_status == "delivered"')\

In [91]:
customers_with_orders.head()

customer_unique_id                          order_id  \
0  861eff4711a542e4b93843c6dd7febb0  00e7ee1b050b8499577073aeb2a297a1   
1  290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2  060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3  259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4  345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   

  order_status order_purchase_timestamp  
0    delivered      2017-05-16 15:05:35  
1    delivered      2018-01-12 20:48:24  
2    delivered      2018-05-19 16:07:45  
3    delivered      2018-03-13 16:06:38  
4    delivered      2018-07-29 09:51:30

In [92]:
# считаем количество дней в каждом месяце
customers_with_orders['month_days'] = customers_with_orders['order_purchase_timestamp'].dt.days_in_month
# считаем количество недель в месяце
customers_with_orders['weeks'] = round(customers_with_orders['month_days']/7,2)
# добавляем столбец месяц-год
customers_with_orders['month_year'] = customers_with_orders['order_purchase_timestamp'].dt.strftime('%Y-%m')

In [93]:
customers_with_orders.head()

customer_unique_id                          order_id  \
0  861eff4711a542e4b93843c6dd7febb0  00e7ee1b050b8499577073aeb2a297a1   
1  290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2  060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3  259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4  345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   

  order_status order_purchase_timestamp  month_days  weeks month_year  
0    delivered      2017-05-16 15:05:35          31   4.43    2017-05  
1    delivered      2018-01-12 20:48:24          31   4.43    2018-01  
2    delivered      2018-05-19 16:07:45          31   4.43    2018-05  
3    delivered      2018-03-13 16:06:38          31   4.43    2018-03  
4    delivered      2018-07-29 09:51:30          31   4.43    2018-07

In [94]:
orders_per_week = customers_with_orders \
    .groupby(['month_year', 'customer_unique_id', 'weeks'], as_index = False)\
    .agg({'order_status' : 'count'})\
    .rename(columns={'order_status': 'orders'})\
    .sort_values('orders',ascending=False)

In [95]:
orders_per_week['order_per_week'] = round(orders_per_week.orders / orders_per_week.weeks, 2)

##### Получаем среднее количество покупок каждого пользователя в неделю

In [96]:
orders_per_week.sort_values('order_per_week',ascending=False)

month_year                customer_unique_id  weeks  orders  \
326      2017-01  12f5d6e1cbf93dafd9dcc19095df0b3d   4.43       6   
51557    2018-02  3e43e6105506432c953e165fb2acf44c   4.00       4   
54531    2018-02  b4e4f24de1e8725b74e4a1f4975116ed   4.00       4   
1986     2017-02  a239b8e2fbce33780f1f1912e2ee5275   4.00       4   
5463     2017-04  25a560b9a6006157838aab1bdbd68624   4.29       4   
...          ...                               ...    ...     ...   
75517    2018-05  d215654948fc3ce8f8010cc12a8c843e   4.43       1   
75518    2018-05  d21fb1e8aee69c608f5c2fb5b6c28cad   4.43       1   
75519    2018-05  d2200863f9f169ddb3a2d6203a8dc007   4.43       1   
75520    2018-05  d22dee9a69c987f183961d4daee38328   4.43       1   
95193    2018-08  fff96bc586f78b1f070da28c4977e810   4.43       1   

       order_per_week  
326              1.35  
51557            1.00  
54531            1.00  
1986             1.00  
5463             0.93  
...               ...  
75517            0.23  
75518            0.23  
75519            0.23  
75520            0.23  
95193            0.23  

[95194 rows x 5 columns]

### 5. Используя pandas, проведи когортный анализ пользователей. В период с января по декабрь выяви когорту с самым высоким retention на 3й месяц

In [175]:
customers_orders.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

    order_approved_at order_delivered_carrier_date  \
0 2017-05-16 15:22:12          2017-05-23 10:47:57   
1 2018-01-12 20:58:32          2018-01-15 17:14:59   
2 2018-05-20 16:19:10          2018-06-11 14:31:00   
3 2018-03-13 17:29:19          2018-03-27 23:22:42   
4 2018-07-29 10:10:09          2018-07-30 15:16:00   

  order_delivered_customer_date order_estimated_delivery_date month_year  
0           2017-05-25 10:35:35                    2017-06-05    2017-05  
1           2018-01-29 12:41:19                    2018-02-06    2018-01  
2           2018-06-14 17:58:51                    2018-06-13    2018-05  
3           2018-03-28 16:04:25                    2018-04-10    2018-03  
4           2018-08-09 20:55:48                    2018-08-15    2018-07

In [201]:
# Добавим столбец с месяцем-годом создания заказа для формирования наших когорт
customers_orders['month_year'] = customers_orders['order_purchase_timestamp'].dt.to_period('M')

In [202]:
cohort = customers_orders[['customer_unique_id', 'order_id', 'order_purchase_timestamp','month_year']]
# и добавим для каждого пользователя колонку с месяцем и годом с формированием первого заказа
cohort['cohort'] = cohort.groupby('customer_unique_id')['order_purchase_timestamp'] \
    .transform('min').dt.to_period('M')

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


Проверим, какие у нас есть периоды

In [203]:
cohort.month_year.value_counts().sort_index()

2016-09       4
2016-10     324
2016-12       1
2017-01     800
2017-02    1780
2017-03    2682
2017-04    2404
2017-05    3700
2017-06    3245
2017-07    4026
2017-08    4331
2017-09    4285
2017-10    4631
2017-11    7544
2017-12    5673
2018-01    7269
2018-02    6728
2018-03    7211
2018-04    6939
2018-05    6873
2018-06    6167
2018-07    6292
2018-08    6512
2018-09      16
2018-10       4
Freq: M, Name: month_year, dtype: int64

Видим, что данные за 2016 и 2018 год у нас не полные, поэтому оставим для выявления retention только 2017 год.
Так же добавим значение "период", который будем показывать нам разницу между первой покупкой и датой заказа

In [204]:
from operator import attrgetter
cohort['period_number'] = (cohort.month_year - cohort.cohort).apply(attrgetter('n'))

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [205]:
cohort = cohort \
    .query("cohort >= '2017-01' and cohort <= '2017-12'")\
    .groupby(['cohort', 'period_number'], as_index=False) \
    .agg({'customer_unique_id':'nunique'}) \
    .rename(columns={'customer_unique_id':'quantity_of_customers'})\
    .pivot(index = 'cohort', columns = 'period_number', values = 'quantity_of_customers')

In [212]:
# Разделим значения в каждом периоде на размер когорты
cohort=(cohort.div(cohort[0], axis='index'))

In [213]:
cohort

period_number   0         1         2         3         4         5   \
cohort                                                                 
2017-01        1.0  0.003927  0.002618  0.001309  0.003927  0.001309   
2017-02        1.0  0.002283  0.002854  0.001142  0.003995  0.001142   
2017-03        1.0  0.004932  0.003794  0.003794  0.003414  0.001517   
2017-04        1.0  0.005952  0.002126  0.001701  0.003401  0.002551   
2017-05        1.0  0.005006  0.005006  0.003893  0.003059  0.003337   
2017-06        1.0  0.004779  0.003504  0.004141  0.002549  0.003823   
2017-07        1.0  0.005136  0.003595  0.002568  0.002825  0.002054   
2017-08        1.0  0.006931  0.003346  0.002629  0.003585  0.005258   
2017-09        1.0  0.006780  0.005327  0.002906  0.004600  0.002179   
2017-10        1.0  0.006935  0.002461  0.000895  0.002237  0.002013   
2017-11        1.0  0.005476  0.003834  0.001780  0.001917  0.001780   
2017-12        1.0  0.002551  0.002734  0.003463  0.002734  0.002005   

period_number        6         7         8         9         10        11  \
cohort                                                                      
2017-01        0.005236  0.001309  0.001309       NaN  0.003927  0.001309   
2017-02        0.002283  0.001712  0.001712  0.002283  0.001142  0.002854   
2017-03        0.001517  0.003035  0.003414  0.000759  0.003794  0.001517   
2017-04        0.003401  0.002976  0.002976  0.001701  0.002551  0.000850   
2017-05        0.004171  0.001669  0.002503  0.003059  0.002503  0.003337   
2017-06        0.003823  0.002230  0.001274  0.002230  0.003186  0.003504   
2017-07        0.003082  0.001027  0.001798  0.002568  0.002311  0.003082   
2017-08        0.002868  0.002629  0.001434  0.001434  0.002390  0.001912   
2017-09        0.002179  0.002421  0.002906  0.001695  0.002663  0.000726   
2017-10        0.002237  0.003579  0.002685  0.002013  0.002013       NaN   
2017-11        0.001095  0.001917  0.001369  0.000548       NaN       NaN   
2017-12        0.001640  0.000547  0.002187       NaN       NaN       NaN   

period_number        12        13        14        15        16        17  \
cohort                                                                      
2017-01        0.007853  0.003927  0.001309  0.001309  0.002618  0.003927   
2017-02        0.001712  0.001712  0.001142  0.000571  0.000571  0.002283   
2017-03        0.002276  0.001138  0.001517  0.002276  0.000759  0.001517   
2017-04        0.000850  0.000425  0.000850  0.000850  0.002126       NaN   
2017-05        0.002503  0.000278  0.001947  0.002503       NaN       NaN   
2017-06        0.001593  0.001274  0.001911       NaN       NaN       NaN   
2017-07        0.001284  0.002568       NaN       NaN       NaN       NaN   
2017-08        0.000956       NaN       NaN       NaN       NaN       NaN   
2017-09             NaN       NaN       NaN       NaN       NaN       NaN   
2017-10             NaN       NaN       NaN       NaN       NaN       NaN   
2017-11             NaN       NaN       NaN       NaN       NaN       NaN   
2017-12             NaN       NaN       NaN       NaN       NaN       NaN   

period_number        19  
cohort                   
2017-01        0.001309  
2017-02             NaN  
2017-03             NaN  
2017-04             NaN  
2017-05             NaN  
2017-06             NaN  
2017-07             NaN  
2017-08             NaN  
2017-09             NaN  
2017-10             NaN  
2017-11             NaN  
2017-12             NaN

In [214]:
retention = (cohort
            .style
            .background_gradient(cmap='Blues')
            .highlight_null('white')
            .format('{:.2%}', na_rep='-'))
retention

##### Таким образом, retention на 3 месяц у нас был самым высоким для июня 2017 года

### 6. Часто для качественного анализа аудитории использую подходы, основанные на сегментации. Используя python, построй RFM-сегментацию пользователей, чтобы качественно оценить свою аудиторию. В кластеризации можешь выбрать следующие метрики: R - время от последней покупки пользователя до текущей даты, F - суммарное количество покупок у пользователя за всё время, M - сумма покупок за всё время. Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров. Пример такого описания: RFM-сегмент 132 (recency=1, frequency=3, monetary=2) имеет границы метрик recency от 130 до 500 дней, frequency от 2 до 5 заказов в неделю, monetary от 1780 до 3560 рублей в неделю.

In [297]:
items_orders

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4       4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
113421  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
113422  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
113423  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
113424  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       29150127e6685892b6eab3eec79f59c7    delivered   
2       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3       951670f92359f4fe4a63112aa7306eba    delivered   
4       6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                  ...          ...   
113420  6760e20addcf0121e9d58f2f1ff14298    delivered   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
113422  fed4434add09a6f332ea398efd656a5c    delivered   
113423  e31ec91cea1ecf97797787471f98a8c2    delivered   
113424  28db69209a75e59f20ccbb5c36a20b90    delivered   

       order_purchase_timestamp   order_approved_at  \
0           2017-05-16 15:05:35 2017-05-16 15:22:12   
1           2018-01-12 20:48:24 2018-01-12 20:58:32   
2           2018-05-19 16:07:45 2018-05-20 16:19:10   
3           2018-03-13 16:06:38 2018-03-13 17:29:19   
4           2018-07-29 09:51:30 2018-07-29 10:10:09   
...                         ...                 ...   
113420      2018-04-07 15:48:17 2018-04-07 16:08:45   
113421      2018-04-04 08:20:22 2018-04-04 08:35:12   
113422      2018-04-08 20:11:50 2018-04-08 20:30:03   
113423      2017-11-03 21:08:33 2017-11-03 21:31:20   
113424      2017-12-19 14:27:23 2017-12-19 18:50:39   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-05-23 10:47:57           2017-05-25 10:35:35   
1               2018-01-15 17:14:59           2018-01-29 12:41:19   
2               2018-06-11 14:31:00           2018-06-14 17:58:51   
3               2018-03-27 23:22:42           2018-03-28 16:04:25   
4               2018-07-30 15:16:00           2018-08-09 20:55:48   
...                             ...                           ...   
113420          2018-04-11 02:08:36           2018-04-13 20:06:37   
113421          2018-04-05 18:42:35           2018-04-11 18:54:45   
113422          2018-04-09 17:52:17           2018-05-09 19:03:15   
113423          2017-11-06 18:24:41           2017-11-16 19:58:39   
113424          2017-12-21 19:17:21           2017-12-26 18:42:36   

       order_estimated_deliv

##### Для RFM-анализа нам потребуются определенные данные о каждом из клиентов:

ID клиента / имя / компания и т.д. - для их идентификации.

Давность (R) как количество дней с момента последней покупки: Сколько дней назад была совершена их последняя покупка? Вычтите дату последней покупки из сегодняшнего дня, чтобы рассчитать значение давности. 1 день назад? 14 дней назад? 500 дней назад?

Частота (F) как общее количество транзакций: Сколько раз клиент совершал покупки в нашем магазине? Например, если кто-то сделал 10 заказов за определенный период времени, его частота равна 10.

Деньги (М) как общая сумма потраченных денег: Сколько $$ (или любой другой валюты расчета) потратил этот клиент? Просто суммируйте деньги от всех транзакций, чтобы получить значение M.

In [298]:
RFM_table = items_orders[['customer_unique_id', 'order_id', 'order_status', 'order_delivered_customer_date', 'product_id', 'price']]

Т.к. изначально мы определили покупкой заказ со статусом delivered, то оставим в таблице только заказы с этим статусом

In [299]:
RFM_table = RFM_table.query("order_status == 'delivered'")[['customer_unique_id', 'order_id', 'order_delivered_customer_date', 'product_id', 'price']]

In [300]:
RFM_table

customer_unique_id                          order_id  \
0       861eff4711a542e4b93843c6dd7febb0  00e7ee1b050b8499577073aeb2a297a1   
1       290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2       060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3       259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4       345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   
...                                  ...                               ...   
113420  1a29b476fee25c95fbafc67c5ac95cf8  6760e20addcf0121e9d58f2f1ff14298   
113421  d52a67c98be1cf6a5c84435bd38d095d  9ec0c8947d973db4f4e8dcf1fbfa8f1b   
113422  e9f50caf99f032f0bf3c55141f019d99  fed4434add09a6f332ea398efd656a5c   
113423  73c2643a0a458b49f58cea58833b192e  e31ec91cea1ecf97797787471f98a8c2   
113424  84732c5050c01db9b23e19ba39899398  28db69209a75e59f20ccbb5c36a20b90   

       order_delivered_customer_date                        product_id   price  
0                2017-05-25 10:35:35  a9516a079e37a9c9c36b9b78b10169e8  124.99  
1                2018-01-29 12:41:19  4aa6014eceb682077f9dc4bffebc05b0  289.00  
2                2018-06-14 17:58:51  bd07b66896d6f1494f5b86251848ced7  139.94  
3                2018-03-28 16:04:25  a5647c44af977b148e0a3a4751a09e2e  149.94  
4                2018-08-09 20:55:48  9391a573abe00141c56e38d84d7d5b3b  230.00  
...                              ...                               ...     ...  
113420           2018-04-13 20:06:37  ccb4503d9d43d245d3b295d0544f988b   74.90  
113421           2018-04-11 18:54:45  9ede6b0570a75a4b9de4f383329f99ee  114.90  
113422           2018-05-09 19:03:15  7a5d2e1e131a860ae7d18f6fffa9d689   37.00  
113423           2017-11-16 19:58:39  f819f0c84a64f02d3a5606ca95edd272  689.00  
113424           2017-12-26 18:42:36  017692475c1c954ff597feda05131d73   13.99  

[110197 rows x 5 columns]

В RFM анализе обычно исследуются данные за определенный период (чаще всего 365 дней). Видим, что покупок в сентябре и октябре не такое большое количество, поэтому будет разумнее исследовать данные с августа 2017 года по август 2018 для формирования полноценной картины.

In [301]:
RFM_table['order_delivered_customer_date'].dt.to_period('M').value_counts().sort_index(ascending=False)

2018-10       3
2018-09      59
2018-08    9385
2018-07    6581
2018-06    7887
2018-05    8166
2018-04    8999
2018-03    7947
2018-02    6623
2018-01    7419
2017-12    8319
2017-11    5412
2017-10    5215
2017-09    4496
2017-08    4900
2017-07    3933
2017-06    3640
2017-05    4201
2017-04    2072
2017-03    2724
2017-02    1565
2017-01     326
2016-12       4
2016-11      72
2016-10     241
Freq: M, Name: order_delivered_customer_date, dtype: int64

In [328]:
# Примем за сегодняшнее число 31 августа 2018 года.
import datetime as dt
NOW = dt.datetime(2018,8,31)
RFM_table_for_year = RFM_table\
    .query("order_delivered_customer_date >= '2017-08-31' and order_delivered_customer_date <= '2018-08-31'")

In [329]:
RFM_table_for_year

customer_unique_id                          order_id  \
1       290c77bc529b7ac935b93aa66c333dc3  29150127e6685892b6eab3eec79f59c7   
2       060e732b5b29e8181a18229c7b0b2b5e  b2059ed67ce144a36e2aa97d2c9e9ad2   
3       259dac757896d24d7702b9acbbff3f3c  951670f92359f4fe4a63112aa7306eba   
4       345ecd01c38d18a9036ed96c73b8d066  6b7d50bd145f6fc7f33cebabd7e49d0f   
5       4c93744516667ad3b8f1fb645a3116a4  5741ea1f91b5fbab2bd2dc653a5b5099   
...                                  ...                               ...   
113420  1a29b476fee25c95fbafc67c5ac95cf8  6760e20addcf0121e9d58f2f1ff14298   
113421  d52a67c98be1cf6a5c84435bd38d095d  9ec0c8947d973db4f4e8dcf1fbfa8f1b   
113422  e9f50caf99f032f0bf3c55141f019d99  fed4434add09a6f332ea398efd656a5c   
113423  73c2643a0a458b49f58cea58833b192e  e31ec91cea1ecf97797787471f98a8c2   
113424  84732c5050c01db9b23e19ba39899398  28db69209a75e59f20ccbb5c36a20b90   

       order_delivered_customer_date                        product_id   price  
1                2018-01-29 12:41:19  4aa6014eceb682077f9dc4bffebc05b0  289.00  
2                2018-06-14 17:58:51  bd07b66896d6f1494f5b86251848ced7  139.94  
3                2018-03-28 16:04:25  a5647c44af977b148e0a3a4751a09e2e  149.94  
4                2018-08-09 20:55:48  9391a573abe00141c56e38d84d7d5b3b  230.00  
5                2017-09-28 17:32:43  0be701e03657109a8a4d5168122777fb  259.90  
...                              ...                               ...     ...  
113420           2018-04-13 20:06:37  ccb4503d9d43d245d3b295d0544f988b   74.90  
113421           2018-04-11 18:54:45  9ede6b0570a75a4b9de4f383329f99ee  114.90  
113422           2018-05-09 19:03:15  7a5d2e1e131a860ae7d18f6fffa9d689   37.00  
113423           2017-11-16 19:58:39  f819f0c84a64f02d3a5606ca95edd272  689.00  
113424           2017-12-26 18:42:36  017692475c1c954ff597feda05131d73   13.99  

[86615 rows x 5 columns]

In [330]:
RFM_table_for_year['order_delivered_customer_date'].dt.to_period('M').value_counts().sort_index(ascending=False)

2018-08    9358
2018-07    6581
2018-06    7887
2018-05    8166
2018-04    8999
2018-03    7947
2018-02    6623
2018-01    7419
2017-12    8319
2017-11    5412
2017-10    5215
2017-09    4496
2017-08     193
Freq: M, Name: order_delivered_customer_date, dtype: int64

In [331]:
RFM_final_table=RFM_table_for_year.groupby('customer_unique_id').agg({'order_delivered_customer_date': lambda x: (NOW - x.max()).days, # Recency
                                                'order_id': lambda x: len(x.unique()), # Frequency
                                                'price': lambda x: x.sum()})    # Monetary 

RFM_final_table['order_delivered_customer_date'] = RFM_final_table['order_delivered_customer_date'].astype(int)

RFM_final_table.rename(columns={'order_delivered_customer_date': 'recency', 
                         'order_id': 'frequency',
                         'price': 'monetary_value'}, inplace=True)

In [332]:
RFM_final_table

recency  frequency  monetary_value
customer_unique_id                                                  
0000366f3b9a7992bf8c76cfdf3221e2      106          1          129.90
0000b849f77a49e4a4ce2b2a4ca5be3f      112          1           18.90
0000f6ccb0745a6a4b88665a16c9f078      302          1           25.99
0004aac84e0df4da2b147fca70cf8255      276          1          180.00
0004bd2a26a76fe21f786e4fbd80607f      145          1          154.00
...                                   ...        ...             ...
fffa431dd3fcdefea4b1777d114144f2      293          1           64.99
fffb09418989a0dbff854a28163e47c6      247          1           58.00
fffbf87b7a1a6fa8b03f081c5f51a201      227          1          149.00
fffea47cd6d3cc0a88bd621562a9d061      233          1           64.89
ffff5962728ec6157033ef9805bacc48      108          1          115.00

[73640 rows x 3 columns]

In [333]:
quantiles = RFM_final_table.quantile(q=[0.25,0.5,0.75])
quantiles

recency  frequency  monetary_value
0.25     80.0        1.0            47.9
0.50    158.0        1.0            89.8
0.75    247.0        1.0           153.0

В нашем случае сегментация по числу покупок является не совсем корректной, т.к. в первом задании мы выяснили, что больше 90% пользователей совершают в нашем магазине только одну покупку, и делить данный показатель на 4 сегмента по квантилям нет смысла.

Разделим сегменты по frequency следующим образом: 

1 - для пользователей, которые совершили только одну покупку 

3 - для пользователей, совершивших больше 1ой покупки

In [334]:
# Converting quantiles to a dictionary, easier to use.
quantiles = quantiles.to_dict()
##  RFM Segmentation ----
RFM_Segment = RFM_final_table.copy()
# Arguments (x = value, p = recency, monetary_value, frequency, k = quartiles dict)
def R_Class(x,p,d):
    if x <= d[p][0.25]:
        return 4
    elif x <= d[p][0.50]:
        return 3
    elif x <= d[p][0.75]: 
        return 2
    else:
        return 1

# Arguments (x = value, p = recency, monetary_value, frequency, k = quartiles dict)
def M_Class(x,p,d):
    if x <= d[p][0.25]:
        return 1
    elif x <= d[p][0.50]:
        return 2
    elif x <= d[p][0.75]: 
        return 3
    else:
        return 4
RFM_Segment['R_Quartile'] = RFM_Segment['recency'].apply(R_Class, args=('recency',quantiles,))
RFM_Segment['F_Segment'] = RFM_Segment['frequency'].apply(lambda x: 1 if x==1 else 3)
RFM_Segment['M_Quartile'] = RFM_Segment['monetary_value'].apply(M_Class, args=('monetary_value',quantiles,))
RFM_Segment['RFMClass'] = RFM_Segment.R_Quartile.map(str) \
                            + RFM_Segment.F_Segment.map(str) \
                            + RFM_Segment.M_Quartile.map(str)

In [335]:
RFM_Segment

recency  frequency  monetary_value  \
customer_unique_id                                                     
0000366f3b9a7992bf8c76cfdf3221e2      106          1          129.90   
0000b849f77a49e4a4ce2b2a4ca5be3f      112          1           18.90   
0000f6ccb0745a6a4b88665a16c9f078      302          1           25.99   
0004aac84e0df4da2b147fca70cf8255      276          1          180.00   
0004bd2a26a76fe21f786e4fbd80607f      145          1          154.00   
...                                   ...        ...             ...   
fffa431dd3fcdefea4b1777d114144f2      293          1           64.99   
fffb09418989a0dbff854a28163e47c6      247          1           58.00   
fffbf87b7a1a6fa8b03f081c5f51a201      227          1          149.00   
fffea47cd6d3cc0a88bd621562a9d061      233          1           64.89   
ffff5962728ec6157033ef9805bacc48      108          1          115.00   

                                  R_Quartile  F_Segment  M_Quartile RFMClass  
customer_unique_id                                                            
0000366f3b9a7992bf8c76cfdf3221e2           3          1           3      313  
0000b849f77a49e4a4ce2b2a4ca5be3f           3          1           1      311  
0000f6ccb0745a6a4b88665a16c9f078           1          1           1      111  
0004aac84e0df4da2b147fca70cf8255           1          1           4      114  
0004bd2a26a76fe21f786e4fbd80607f           3          1           4      314  
...                                      ...        ...         ...      ...  
fffa431dd3fcdefea4b1777d114144f2           1          1           2      112  
fffb09418989a0dbff854a28163e47c6           2          1           2      212  
fffbf87b7a1a6fa8b03f081c5f51a201           2          1           3      213  
fffea47cd6d3cc0a88bd621562a9d061           2          1           2      212  
ffff5962728ec6157033ef9805bacc48           3          1           3      313  

[73640 rows x 7 columns]

In [336]:
# Посмотрим, как разбились наши сегменты
RFM_Segment.RFMClass.value_counts()

211    4784
112    4771
411    4663
313    4611
212    4606
413    4561
412    4517
111    4478
213    4459
311    4430
114    4413
314    4367
113    4367
414    4343
312    4278
214    4133
434     336
334     326
234     275
134     193
233     116
433     115
333     102
133      97
332      67
432      66
132      57
232      47
331      23
431      20
231      10
131       9
Name: RFMClass, dtype: int64

In [337]:
# ТОП-5 наших самых потерянных пользователй - совершали одну минимальную покупку 364 дня назад
RFM_Segment[RFM_Segment['RFMClass']=='111'].sort_values('recency',ascending=False).head(5)

recency  frequency  monetary_value  \
customer_unique_id                                                     
118be57d1b7de439453a0c6625ad7934      364          1           29.90   
198ab28ceee7a2311444e5c8c9109558      364          1           47.90   
e47efe0febf5cbec6ea5456257f96e5a      364          1           19.90   
fe59d5878cd80080edbd29b5a0a4e1cf      364          1            9.99   
7824884ec60635edc70c6706b71488b8      364          1           40.00   

                                  R_Quartile  F_Segment  M_Quartile RFMClass  
customer_unique_id                                                            
118be57d1b7de439453a0c6625ad7934           1          1           1      111  
198ab28ceee7a2311444e5c8c9109558           1          1           1      111  
e47efe0febf5cbec6ea5456257f96e5a           1          1           1      111  
fe59d5878cd80080edbd29b5a0a4e1cf           1          1           1      111  
7824884ec60635edc70c6706b71488b8           1          1           1      111

In [340]:
# ТОП-5 наших лояльных клиентов - совершили несколько покупок на сумму выше среднего
RFM_Segment[RFM_Segment['F_Segment'] == 3 ].sort_values('monetary_value', ascending=False).head(5)

recency  frequency  monetary_value  \
customer_unique_id                                                     
c8460e4251689ba205045f3ea17884a1        9          4         4080.00   
d77aa95864ae5b42160937615628723a      359          2         2400.00   
7b0eaf68a16e4808e5388c67345033c9       85          2         2238.42   
6ddbc64bd04d40f7768ff088d94cbeb8      136          2         2108.00   
73601b1eec55943e90ce8d61253d5c09      293          2         2009.98   

                                  R_Quartile  F_Segment  M_Quartile RFMClass  
customer_unique_id                                                            
c8460e4251689ba205045f3ea17884a1           4          3           4      434  
d77aa95864ae5b42160937615628723a           1          3           4      134  
7b0eaf68a16e4808e5388c67345033c9           3          3           4      334  
6ddbc64bd04d40f7768ff088d94cbeb8           3          3           4      334  
73601b1eec55943e90ce8d61253d5c09           1          3           4      134

##### Как мы видим среди наших лояльных клиентов есть такие, которые уже давно не совершали покупок. Нужно задуматься о способах их возвращения: рассылка со спецпредложениями, бонусами, акциями